### PR TITLE
UniqueType-i-fying UniqueTriggerActivation

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
Begs the question - guideline to group and comment UniqueTypes... But somehow there's something amiss in that diff.

### CODE REVIEW

**Comment by SomeTroglodyte** [comment]
Ah, #5389 already fixed the diff I was missing here, so that's OK then.

This is intended as groundwork, I did my best to not change behaviour in any way.

File: core/src/com/unciv/models/ruleset/unique/UniqueType.kt [hunk 1]
```diff
@@ -220,6 +220,43 @@ enum class UniqueType(val text:String, vararg targets: UniqueTarget) {
     // tile conditionals
     ConditionalNeighborTiles("with [amount] to [amount] neighboring [tileFilter] tiles", UniqueTarget.Conditional),
     ConditionalNeighborTilesAnd("with [amount] to [amount] neighboring [tileFilter] [tileFilter] tiles", UniqueTarget.Conditional),
+
+    ///////////////////////////////////////// TRIGGERED ONE-TIME /////////////////////////////////////////
+
+    OneTimeFreeUnit("Free [baseUnitFilter] appears", UniqueTarget.Global),  // used in Policies, Buildings
+    OneTimeAmountFreeUnits("[amount] free [baseUnitFilter] units appear", UniqueTarget.Global), // used in Buildings
+    OneTimeFreeUnitRuins("Free [] found in the ruins", UniqueTarget.Ruins), // Differs from "Free [] appears" in that it spawns near the ruins instead of in a city
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueType.kt:228]: Missing

File: core/src/com/unciv/models/ruleset/unique/UniqueType.kt [hunk 2]
```diff
@@ -220,6 +220,43 @@ enum class UniqueType(val text:String, vararg targets: UniqueTarget) {
...
+    OneTimeFreePolicy("Free Social Policy", UniqueTarget.Global), // used in Buildings
+    OneTimeAmountFreePolicies("[amount] Free Social Policies", UniqueTarget.Global),  // Not used in Vanilla
+    OneTimeEnterGoldenAge("Empire enters golden age", UniqueTarget.Global),  // used in Policies, Buildings
+    OneTimeFreeGreatPerson("Free Great Person", UniqueTarget.Global),  // used in Policies, Buildings
+    OneTimeGainPopulation("[amount] population [cityFilter]", UniqueTarget.Global),  // used in CN tower
+    OneTimeGainPopulationRandomCity("[] population in a random city", UniqueTarget.Ruins),
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueType.kt:234]: Missing

File: core/src/com/unciv/models/ruleset/unique/UniqueType.kt [hunk 3]
```diff
@@ -220,6 +220,43 @@ enum class UniqueType(val text:String, vararg targets: UniqueTarget) {
...
+    OneTimeRevealEntireMap("Reveals the entire map", UniqueTarget.Global),  // used in tech
+    OneTimeGainStat("Gain [amount] [stat]", UniqueTarget.Ruins),
+    OneTimeGainStatRange("Gain [amount]-[amount] [stat]", UniqueTarget.Ruins),
+    OneTimeGainPantheon("Gain enough Faith for a Pantheon", UniqueTarget.Ruins),
+    OneTimeGainProphet("Gain enough Faith for [amount]% of a Great Prophet", UniqueTarget.Ruins),
+    OneTimeRevealSpecificMapTiles("Reveal up to [amountOrAll] [tileFilter] within a [amount] tile radius", UniqueTarget.Ruins),
+    OneTimeRevealCrudeMap("From a randomly chosen tile [amount] tiles away from the ruins, reveal tiles up to [amount] tiles away with [amount]% chance", UniqueTarget.Ruins),
+    OneTimeTriggerVoting("Triggers voting for the Diplomatic Victory", UniqueTarget.Global),  // used in Building
+
+    OneTimeUnitHeal("Heal this unit by [] HP", UniqueTarget.Promotion),
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueType.kt:247]: We should probably as a test for [] in unique type texts

**Comment by SomeTroglodyte** [comment]
OK - another thought: If we rely on placeholderText or (placeholderText+target) to be unique - ahem - a unit test for that?

File: core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt [hunk 4]
```diff
@@ -15,6 +15,10 @@ enum class UniqueParameterType(val parameterName:String) {
             else null
         }
     },
+    NumberOrAll("amountOrAll") {
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: Denied, this is a clear "amount". "Reveal up to All" throws human parse errors

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]:
```json
    {
        "name": "reveal nearby Barbarian camps",
        "notification": "You find evidence of Barbarian activity. Nearby Barbarian camps are revealed!",
        "uniques": ["Reveal up to [All] [Barbarian encampment] within a [10] tile radius"]
    },
```
Not mine

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: > throws human parse errors

:grin:  and the human throws his breakfast in to the exception bucket (Mr Creosote)

Comment (yairm210) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: Eeww yukkk
"amount/param" then, and please add a todo to that

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: Wait - the "/" is parsed as alternate types? OK, forgot, that would obviously obviate the need for _that_ kludge type

Comment (yairm210) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: That is, this shouldn't be an enum value

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: No "/" isn't parsed - didn't someone code that? Thought I saw it earlier. Do we want? Right in UniqueType.init?

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: And that init could really use that safeValueOf generic I got in stash - time to unstash?
```kotlin
inline fun <reified T: Enum<T>> enumValueOfOrNull(name: String?): T? {
    if (name == null) return null
    return try {
        enumValueOf<T>(name)
    } catch (ex: IllegalArgumentException) {
        println("Could not parse\"$name\" as ${T::class.java.simpleName}.")
        null
    }
}
```
Each enum then can do a proxy easily as one-liner...

Comment (xlenstra) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: Hey, that looks like code I've written :p
On second thought, it's probably best to split this out into two uniques?

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: Our unit tests do not balk at parameter type violations as defined in UniqueParameterType.getErrorSeverity at all - uh, where were we? That still on the to-do list? Anyway, separating that all barb camp (plural is another readability problem) ruin reward - not this PR?

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt:18]: OK, `checkModLinks` does not check the _ruin rewards_ and I was only looking for a test failure manipulation those... Yet another PR.

Comment (SomeTroglodyte) [reply core/src/com/unciv/models/ruleset/unique/UniqueParameterType.kt]: Used a simpler safeValueOf in that latest push, also implements "/" for autoparsed placeholder params, so now those lists can have size>1

**Comment by SomeTroglodyte** [comment]
Of course Unciv fails tests once `checkModLinks` covers _all_ unique-containing ruleset objects and once `@Test fun uniqueTypesHaveNoUnknownParameters()` is in there... On it.

**Comment by SomeTroglodyte** [comment]
```
Error: Air Targeting I requires promotion Dogfighting I which does not exist!
Error: Sortie requires promotion Dogfighting II which does not exist!
Warning: Fertility Rites's unique "[+10]% growth [in this city]" cannot be put on this type of object!
Warning: Guruship's unique "[+2 Production] <if this city has at least [1] specialists>" cannot be put on this type of object!
Warning: Swords into Ploughshares's unique "[+15]% growth [in this city] <when not at war>" cannot be put on this type of object!
```
```
StatPercentBonus param[1] type is unknown
ConsumesResources param[1] type is unknown
ProvidesResources param[1] type is unknown
FreeExtraBeliefs param[1] type is unknown
FreeExtraBeliefs param[2] type is unknown
FreeExtraAnyBeliefs param[1] type is unknown
StrengthPlusVs param[1] type is unknown
StrengthMinVs param[1] type is unknown
NaturalWonderConvertNeighbors param[0] type is unknown
NaturalWonderConvertNeighborsExcept param[0] type is unknown
NaturalWonderConvertNeighborsExcept param[1] type is unknown
```
That's not too bad... The second test could pull out the actually used names though.

**Pull request merged**